### ISSUE TITLE
Add tooling or guidance on how a Spicy parser with error recovery should deal with huge of packet loss

### ISSUE THREAD

**Issue by bbannier** [zeek/zeek#3779]
Spicy parser can provide [error recovery](https://docs.zeek.org/projects/spicy/en/latest/programming/parsing.html#error-recovery) so that they can resynchronize after parse errors (probably most common: after parse errors triggered by gaps due to packet loss). During error recovery the Spicy parser will enter a search mode where it searches the input stream for a pattern to synchronize on.

This process works, but is not free since the search might be (much) more expensive than parsing, especially if synchronization fails repeatedly due to more gaps before a match is found, and users reported that parsers spend most of their CPU time in recovery if there is high packet loss. We currently have the following ways to deal with that:

- use `%synchronize-at`/`%synchronize-after` to use a dedicated and possible more optimized search pattern
- zeek/spicy#1673: allow to use a custom function to find a synchronization points

If packet loss is high enough these will ultimately still fail (similar argument why even a more optimized search impl with zeek/spicy#1133 would at a certain point also be insufficient).

We should provide or at least document an approach for parsers to opt-out of further parsing of a connection if packet loss is too high. An automated approach could look at some packet loss metric (which one?) and then deactivate the parser if it exceeds a threshold (likely: configurable since I'd expect the trade-off between visibility and throughput to be site-specific). Alternatively we could also document an approach on how to implement such a mitigation (e.g., inspect stats counters and manually perform the deactivation). It could be that either of this requires additional Spicy-side hooks or metrics.

**Comment by rsmmr** [comment]
Here's an idea: we could add statistics to the current input `stream` tracking number of data chunks vs gaps. Then, during synchronization, an analyzer could retrieve those stats and decide if there's just too many gaps for it to reasonably continue and disable itself by calling  `zeek::skip_input()`. To retrieve the stats(), we'd introduce some built-in function. As the decision point where to check the stats, one could use one of the existing [synchronization hooks](https://docs.zeek.org/projects/spicy/en/latest/programming/parsing.html#error-recovery-hooks). And/or, I think it would be reasonable to extend the existing [on_gap hook](https://docs.zeek.org/projects/spicy/en/latest/programming/language/types.html#method-%on_gap) for sinks to trigger for normal input stream gaps as well. @Mohan-Dhawan, what do you think?

**Comment by Mohan-Dhawan** [comment]
@rsmmr Use of stats to guide synchronization would be perfect. Just one query, once disabled the analyzer cannot be re-enabled for that input, right?

**Comment by rsmmr** [comment]
> Just one query, once disabled the analyzer cannot be re-enabled for that input, right?

That's correct (which seems reasonable to me, how would you make that decision to reenable?)
